### PR TITLE
Feature/sink canal protocol

### PR DESCRIPTION
**Pull request by oscar-zyh**
### What problem does this PR solve? 
support canal protocol in mqSink 
https://github.com/pingcap/ticdc/issues/607

### What is changed and how it works?
implements the EventBatchEncoder interface by CanalEventBatchEncoder

Tests 

 - Manual test (add detailed scripts or steps below)

### Release note

- support canal protocol in mq sink

### CODE REVIEW

File: cdc/sink/codec/interface.go [hunk 1]
```diff
@@ -44,3 +49,37 @@ type EventBatchDecoder interface {
...
+
+// NewEventBatchEncoder returns a function of creating an EventBatchEncoder
+func NewEventBatchEncoder(p Protocol) func() EventBatchEncoder {
+	switch p {
+	case ProtocolDefault:
+		return NewJSONEventBatchEncoder
+	case ProtocolCanal:
+		return NewCanalEventBatchEncoder
+	default:
+		return NewJSONEventBatchEncoder
```
Comment (leoppro) [inline cdc/sink/codec/interface.go:83]: I think we'd better add a warning to tell the user that the protocol of mq sink is wrong.

File: cdc/sink/codec/interface.go [hunk 2]
```diff
@@ -44,3 +49,37 @@ type EventBatchDecoder interface {
...
+// Enum types of the Protocol
+const (
+	ProtocolDefault Protocol = iota
+	ProtocolCanal
+)
+
+// FromString converts the protocol from string to Protocol enum type
+func (p *Protocol) FromString(protocol string) {
+	switch strings.ToLower(protocol) {
+	case "default":
```
Comment (leoppro) [inline cdc/sink/codec/interface.go:65]: I think `json` is more appropriate, and rename `ProtocolDefault` to `ProtocolJson`
```suggestion
	case "json":
```

File: cdc/sink/codec/canal.go [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+	if newSize > oldSize {
+		// resize packet body slice
+		d.packet.Body = append(d.packet.Body, make([]byte, newSize-oldSize)...)
+	}
+	_, err := d.messages.MarshalToSizedBuffer(d.packet.Body[:newSize])
+	return err
+}
+
+// NewCanalEventBatchEncoder creates a new CanalEventBatchEncoder.
+func NewCanalEventBatchEncoder() EventBatchEncoder {
```
Comment (leoppro) [inline cdc/sink/codec/canal.go:348]: could you add a unit test for canal event batch encoder?

File: cdc/sink/codec/interface.go [hunk 4]
```diff
@@ -13,7 +13,12 @@
 
 package codec
 
-import "github.com/pingcap/ticdc/cdc/model"
+import (
+	"github.com/pingcap/log"
+	"github.com/pingcap/ticdc/cdc/model"
+	"go.uber.org/zap"
+	"strings"
```
Comment (amyangfei) [inline cdc/sink/codec/interface.go:20]: recommend organizing import as 
```
standard lib

third-party lib
```

File: cdc/sink/codec/interface.go (hunk 2 above)
Comment (amyangfei) [reply cdc/sink/codec/interface.go:65]: Maybe `json` is not accurate here, in fact the json encoder is a batch implement of TiCDC open protocol.

File: cdc/sink/codec/interface.go (hunk 4 above)
Comment (leoppro) [reply cdc/sink/codec/interface.go:20]: https://github.com/pingcap/ticdc/pull/649#discussion_r439985220

any suggestions about this?

File: proto/CanalProtocol.proto [hunk 5]
```diff
@@ -0,0 +1,154 @@
+syntax = "proto3";
```
Comment (overvenus) [inline proto/CanalProtocol.proto:1]: Could you add Alibaba Apache-2.0 license header?

File: proto/EntryProtocol.proto [hunk 6]
```diff
@@ -0,0 +1,230 @@
+syntax = "proto3";
```
Comment (overvenus) [inline proto/EntryProtocol.proto:1]: Could you add Alibaba Apache-2.0 license header?

File: cdc/sink/codec/interface.go (hunk 4 above)
Comment (oscar-zyh) [reply cdc/sink/codec/interface.go]: fixed in https://github.com/pingcap/ticdc/pull/649/commits/c6079f5a8826846fd9c6330eda302b80b432c905

File: cdc/sink/codec/canal.go (hunk 3 above)
Comment (oscar-zyh) [reply cdc/sink/codec/canal.go:348]: unit tests added in https://github.com/pingcap/ticdc/pull/649/commits/c6079f5a8826846fd9c6330eda302b80b432c905

File: cdc/sink/codec/interface.go (hunk 1 above)
Comment (oscar-zyh) [reply cdc/sink/codec/interface.go:85]: fixed in https://github.com/pingcap/ticdc/pull/649/commits/c6079f5a8826846fd9c6330eda302b80b432c905

File: proto/CanalProtocol.proto (hunk 5 above)
Comment (oscar-zyh) [reply proto/CanalProtocol.proto:17]: added in https://github.com/pingcap/ticdc/pull/649/commits/01a5a3ef9510784039fa58e35da42e0b142c81a9

File: proto/EntryProtocol.proto (hunk 6 above)
Comment (oscar-zyh) [reply proto/EntryProtocol.proto:17]: added in https://github.com/pingcap/ticdc/pull/649/commits/01a5a3ef9510784039fa58e35da42e0b142c81a9

**Pull request merged**